primary/src/metrics.rs

```diff
@@ -0,0 +1,128 @@
+use std::sync::Once;
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::EndpointMetrics;
+use prometheus_exporter::prometheus::{
+    default_registry, register_histogram_vec_with_registry, register_int_counter_vec_with_registry,
+    HistogramVec, IntCounterVec, Registry,
+};
+
+/// The prefix for the primary node metrics
+pub(crate) const PRIMARY_METRICS_PREFIX: &str = "narwhal_primary";
+
+#[derive(Clone)]
+pub(crate) struct Metrics {
+    pub(crate) endpoint_metrics: Option<EndpointMetrics>,
+    pub(crate) node_metrics: Option<PrimaryMetrics>,
+}
+
+static mut METRICS: Metrics = Metrics {
+    endpoint_metrics: None,
+    node_metrics: None,
+};
+static INIT: Once = Once::new();
+
+/// Initialises the metrics. Should be called only once when the primary
+/// node is initialised, otherwise it will lead to erroneously creating
+/// multiple registries.
+pub(crate) fn initialise_metrics() -> Metrics {
+    unsafe {
+        INIT.call_once(|| {
+            let metrics_registry = default_registry();
```

Comment:
Ideally I would like to create a [custom Registry](https://github.com/tikv/rust-prometheus/blob/935d968333102c16a954a87295ea8e6c355df605/src/registry.rs#L232) here so I could define more properly the metrics `prefix` and standard labels (e.x the primary's name) so those could be applied on any metric. Unfortunately the [prometheus-exporter](https://github.com/AlexanderThaller/prometheus_exporter) is using the default registry to derive the metrics and there is not way to customise it - which is unfortunate and I am thinking that we could probably fork the project and just amend our own - after all is not a super bug lib. Unless we create our own custom server solution which we'll of course make it configurable anyways

Reply:
Done 👍  I've also allowed to pass the registry as part of the [Node::spawn_primary](https://github.com/MystenLabs/narwhal/blob/65993a63ffb23ddd6e0cc7253417682930bbf64f/node/src/lib.rs#L92) and extracted the spin up of the prometheus metrics outside of it as the primary is [spawned in SUI as well ](https://github.com/MystenLabs/sui/blob/e33bda526ab2f6714258054bf30c4effd9df05d4/crates/sui-core/src/authority_server.rs#L174) and this would lead to:
1. Spinning up an unnecessary web server to export prometheus metrics under a different port + endpoint
2. Metrics would be reported under a different registry, which probably don't want to

now you can pass the same registry to initialise in SUI node down to Narwhal nodes as well.

Reply:
That's awesome! allows us to register the narwhal metrics from a sui registry